src/CentrifugeRouter.sol

```diff
@@ -64,30 +60,90 @@ contract CentrifugeRouter is Auth, Multicall, ICentrifugeRouter {
+        lockedRequests[controller][vault] = 0;
+        IERC7540(vault).requestDeposit(lockedRequest, controller, address(this));
+        emit ExecuteLockedDepositRequest(vault, controller);
     }
 
     /// @inheritdoc ICentrifugeRouter
-    function claimDeposit(address vault, address user) external {
-        uint256 maxDeposit = IERC7540(vault).maxDeposit(user);
-        IERC7540(vault).deposit(maxDeposit, user, user);
+    function claimDeposit(address vault, address receiver, address controller) external {
+        uint256 maxDeposit = IERC7540(vault).maxDeposit(controller);
+        IERC7540(vault).deposit(maxDeposit, receiver, controller);
     }
 
     // --- Redeem ---
     /// @inheritdoc ICentrifugeRouter
-    function requestRedeem(address vault, uint256 amount) external {
-        IERC7540(vault).requestRedeem(amount, msg.sender, msg.sender);
+    function requestRedeem(address vault, uint256 amount, address controller, address owner) external {
+        IERC7540(vault).requestRedeem(amount, controller, owner);
+    }
+
+    /// @inheritdoc ICentrifugeRouter
+    function claimRedeem(address vault, address receiver, address controller) external {
+        uint256 maxRedeem = IERC7540(vault).maxRedeem(controller);
+        IERC7540(vault).redeem(maxRedeem, receiver, controller);
+    }
+
+    // --- ERC20 approval ---
+    /// @inheritdoc ICentrifugeRouter
+    function approveMax(address token, address spender) external {
+        if (IERC20(token).allowance(address(this), spender) == 0) {
+            SafeTransferLib.safeApprove(token, spender, type(uint256).max);
+        }
     }
 
     /// @inheritdoc ICentrifugeRouter
-    function claimRedeem(address vault, address user) external {
-        uint256 maxRedeem = IERC7540(vault).maxRedeem(user);
-        IERC7540(vault).redeem(maxRedeem, user, user);
+    function permit(address asset, uint256 assets, uint256 deadline, uint8 v, bytes32 r, bytes32 s) external {
+        try IERC20Permit(asset).permit(msg.sender, address(this), assets, deadline, v, r, s) {} catch {}
+    }
+
+    // --- ERC20 wrapping ---
+    /// @dev requires calling approveMax(underlying, wrapper) before
+    function wrap(address wrapper, uint256 amount, address receiver) external {
+        address underlying = IERC20Wrapper(wrapper).underlying();
+
+        amount = MathLib.min(amount, IERC20(underlying).balanceOf(msg.sender));
+        require(amount != 0, "CentrifugeRouter/zero-balance");
+        SafeTransferLib.safeTransferFrom(underlying, msg.sender, address(this), amount);
+
+        require(IERC20Wrapper(wrapper).depositFor(receiver, amount), "CentrifugeRouter/deposit-for-failed");
+    }
+
+    function unwrap(address wrapper, uint256 amount, address receiver) external {
+        require(receiver != address(0), "CentrifugeRouter/zero-address");
+        amount = MathLib.min(amount, IERC20(wrapper).balanceOf(address(this)));
```

Comment:
Is this  good UX ? I ask the contract to give me 100 units but it sends back to me 50 only because it's all it had? If I have no underlying knowledge about the implementation I would be frightened.

Reply:
I basically copied this from Morpho: https://github.com/morpho-org/morpho-blue-bundlers/blob/main/src/ERC20WrapperBundler.sol#L33

But I see your point, let me think about it!

Reply:
I realized why it is the way it is, and it makes sense: if you are chaining multiple actions, you might not know the exact output of some of the actions. So with this, you can overestimate the output of 1 action and provide that as the input to a later action. That wouldn't work if you force exact amounts.

Reply:
Indeed it makes sense the more you think about it :)